clusterloader2/pkg/measurement/common/service_creation_latency.go

```diff
@@ -0,0 +1,317 @@
+// String returns a string representation of the metric.
+func (s *serviceCreationLatencyMeasurement) String() string {
+	return serviceCreationLatencyName + ": " + s.selector.String()
+}
+
+func (s *serviceCreationLatencyMeasurement) start() error {
+	if s.isRunning {
+		klog.Infof("%s: wait for controlled pods measurement already running", s)
+		return nil
+	}
+	klog.Infof("%s: starting wait for controlled pods measurement...", s)
+
+	s.isRunning = true
+	s.stopCh = make(chan struct{})
+
+	i := informer.NewInformer(
+		s.client,
+		"services",
+		s.selector,
+		func(oldObj, newObj interface{}) {
+			f := func() {
+				s.handleObject(oldObj, newObj)
+			}
+			s.queue.Add(&f)
+		},
+	)
+	return informer.StartAndSync(i, s.stopCh, informerSyncTimeout)
+}
+
+func (s *serviceCreationLatencyMeasurement) waitForReady() error {
+	for {
+		created, ipAssigned, reachable := 0, 0, 0
+		select {
+		case <-time.After(s.waitTimeout):
+			return fmt.Errorf("waiting for services timeout")
+		case <-time.After(defaultCheckInterval):
+			func() {
+				reachable = s.creationTimes.Count(reachabilityPhase)
+				ipAssigned = s.creationTimes.Count(ipAssigningPhase)
+				created = s.creationTimes.Count(creatingPhase)
+
+			}()
+			klog.Infof("%s: %d created, %d ipAssigned, %d reachable", s, created, ipAssigned, reachable)
+			if created == reachable {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering service created latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	serviceCreationLatency := s.creationTimes.CalculateTransitionsLatency(map[string]measurementutil.Transition{
+		"create_to_assigned": {
+			From: creatingPhase,
+			To:   ipAssigningPhase,
+		},
+		"assigned_to_available": {
+			From: ipAssigningPhase,
+			To:   reachabilityPhase,
+		},
+		"create_to_available": {
+			From: creatingPhase,
+			To:   reachabilityPhase,
+		},
+	})
+
+	content, err := util.PrettyPrintJSON(measurementutil.LatencyMapToPerfData(serviceCreationLatency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, err
+}
+
+func (s *serviceCreationLatencyMeasurement) handleObject(oldObj, newObj interface{}) {
+	var oldService *corev1.Service
+	var newService *corev1.Service
+	var ok bool
+	oldService, ok = oldObj.(*corev1.Service)
+	if oldObj != nil && !ok {
+		klog.Errorf("%s: uncastable old object: %v", s, oldObj)
+		return
+	}
+	newService, ok = newObj.(*corev1.Service)
+	if newObj != nil && !ok {
+		klog.Errorf("%s: uncastable new object: %v", s, newObj)
+		return
+	}
+	isEqual := oldService != nil &&
+		newService != nil &&
+		equality.Semantic.DeepEqual(oldService.Spec, newService.Spec) &&
+		equality.Semantic.DeepEqual(oldService.Status, newService.Status)
+	if isEqual {
+		return
+	}
+
+	if !s.isRunning {
```

Comment:
This isn't thread-safe - you're accessing this value and can modify it when calling Dispose()
This isn't specific to this measurement - it's a problem for many existing measurements (and it works only because we're cleaning ourselves in test, so nothing is happening when Dispose is called).

Please file an issue for this.

Reply:
Done.